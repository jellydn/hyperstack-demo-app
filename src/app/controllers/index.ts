// naming and imports: name and import as you see fit
import Auth from './auth'
import { NotesController } from './notes'

const controllers = []
controllers.push(Auth)
controllers.push(NotesController)

export const AppController = {
  controllers,
}
