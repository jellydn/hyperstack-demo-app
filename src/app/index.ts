import type { Context } from 'hyperstack'
import { context } from 'hyperstack'
import type * as Models from './models'
import type * as Mailers from './mailers'

export const appContext = context as Context<typeof Models, typeof Mailers>
