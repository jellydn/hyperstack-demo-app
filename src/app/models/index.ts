export { Note } from './note'
export { User } from './user'
