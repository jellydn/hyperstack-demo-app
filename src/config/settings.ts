// general purpose settings, with no infrastructure loading, no assumptions. keep it simple.
import path from 'path'
export const root = path.join(__dirname, '..')
