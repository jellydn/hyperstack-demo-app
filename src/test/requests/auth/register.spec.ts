import { test } from '@hyperstackjs/testing'
import { root } from '../../../config/settings'
import { appContext } from '../../../app'
const {
  requests,
  matchers: { matchRequestWithSnapshot, matchDeliveriesWithSnapshot },
  serializers: { redactAndExpectMatch, baseMailerSerialize },
} = test(root)

const serializer = redactAndExpectMatch({
  'body.token': /^eyJh.+/,
})

const verifyTokenExpr = /verifyToken=[a-z0-9]{64}/g
const serializeVerifyEmails = redactAndExpectMatch(
  {
    'contents.1': verifyTokenExpr,
    'contents.2': verifyTokenExpr,
    'originalMessage.html': verifyTokenExpr,
    'originalMessage.text': verifyTokenExpr,
  },
  baseMailerSerialize
)

const serializeUser = redactAndExpectMatch({
  pid: /.+-.+-.+-.+-.+/,
})
jest.setTimeout(10000)
describe('requests', () => {
  describe('/auth/register', () => {
    requests('should register and send welcome email', async (request) => {
      const { User } = appContext.models()
      await matchRequestWithSnapshot(
        200,
        request().post('/auth/register').send({
          username: 'evh@example.com',
          password: 'mypass-should-login',
          name: 'Eddie Van Halen',
        }),
        { serializer }
      )
      const user = await User.findOne({
        where: { username: 'evh@example.com' },
      })
      if (!user) {
        throw new Error('user is null')
      }
      expect(user.emailVerificationToken).toMatch(/.{3,}/)
      expect(user.emailVerificationSentAt).toBeTruthy()

      expect(
        new Date().getTime() - user.emailVerificationSentAt!.getTime()
      ).toBeWithin(0, 5000)

      expect(serializeUser(user)).toMatchSnapshot()
      matchDeliveriesWithSnapshot(1, { serializer: serializeVerifyEmails })
    })

    requests('should validate', async (request) => {
      const { User } = appContext.models()
      await User.createWithPassword({
        username: 'evh@example.com',
        password: 'mypass-should-login',
        name: 'Eddie Van Halen',
      })
      await matchRequestWithSnapshot(
        400,
        request().post('/auth/register').send({
          username: 'no-such-user@example.com',
          password: 'mypass-should-login',
        })
      )
      await matchRequestWithSnapshot(
        400,
        request().post('/auth/register').send({
          username: 'this-is-not-email',
          password: 'mypass-should-login',
          name: 'Eddie Van Halen',
        })
      )
      // password too short
      await matchRequestWithSnapshot(
        400,
        request().post('/auth/register').send({
          username: 'evh@example.com',
          password: 'boo',
          name: 'Eddie Van Halen',
        })
      )
      // user exists
      await matchRequestWithSnapshot(
        400,
        request().post('/auth/register').send({
          username: 'evh@example.com',
          password: 'mypass-should-login',
          name: 'Eddie Van Halen',
        })
      )
    })

    requests('should verify with secret token', async (request) => {
      const { User } = appContext.models()
      const user = await User.createWithPassword({
        username: 'evh@example.com',
        password: 'mypass-should-login',
        name: 'Eddie Van Halen',
      })
      const verifyToken = user.emailVerificationToken

      await matchRequestWithSnapshot(
        200,
        request().get(`/auth/verify?verifyToken=${verifyToken}`),
        { serializer }
      )
      await user.reload()
      expect(new Date().getTime() - user.emailVerifiedAt!.getTime()).toBeWithin(
        0,
        500
      )
      expect(user.emailVerificationToken).toBeNull()
      expect(user.emailVerificationSentAt).toBeNull()

      // no email bloopers
      matchDeliveriesWithSnapshot(0)
    })
  })
})
