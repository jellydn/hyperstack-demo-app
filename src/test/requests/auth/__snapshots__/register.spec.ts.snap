// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requests /auth/register should register and send welcome email 1`] = `
{
  "body": {
    "token": "test-redacted",
  },
  "headers": {
    "connection": "close",
    "content-length": "252",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`requests /auth/register should register and send welcome email 2`] = `
{
  "name": "Eddie Van Halen",
  "pid": "test-redacted",
  "username": "evh@example.com",
}
`;

exports[`requests /auth/register should register and send welcome email 3`] = `
[
  {
    "content": undefined,
    "contentType": "multipart/alternative",
    "contents": [
      undefined,
      "Welcome Eddie Van Halen, you can now log in.
  Verify your account with the link below:

  http://localhost/verify?test-redacted
",
      "<html>

<body>
  You can <a href="http://localhost:/verify?test-redacted">verify your account</a>
</body>

</html>
",
    ],
    "envelope": {
      "from": "no-reply@elle-postage.example.com",
      "to": [
        "evh@example.com",
      ],
    },
    "from": "no-reply@elle-postage.example.com",
    "messageId": "test-redacted",
    "originalMessage": {
      "attachments": [],
      "from": "Elle Postage <no-reply@elle-postage.example.com>",
      "html": "<html>

<body>
  You can <a href="http://localhost:/verify?test-redacted">verify your account</a>
</body>

</html>
",
      "subject": "Welcome Eddie Van Halen",
      "text": "Welcome Eddie Van Halen, you can now log in.
  Verify your account with the link below:

  http://localhost/verify?test-redacted
",
      "to": "evh@example.com",
    },
    "response": "test-redacted",
    "subject": "Welcome Eddie Van Halen",
    "to": [
      "evh@example.com",
    ],
  },
]
`;

exports[`requests /auth/register should validate 1`] = `
{
  "body": {
    "error": "validation",
    "validationErrors": [
      {
        "code": "invalid_type",
        "expected": "string",
        "message": "Required",
        "path": [
          "name",
        ],
        "received": "undefined",
      },
    ],
  },
  "headers": {
    "connection": "close",
    "content-length": "147",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`requests /auth/register should validate 2`] = `
{
  "body": {
    "error": "validation",
    "validationErrors": [
      {
        "code": "invalid_string",
        "message": "Invalid email",
        "path": [
          "username",
        ],
        "validation": "email",
      },
    ],
  },
  "headers": {
    "connection": "close",
    "content-length": "136",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`requests /auth/register should validate 3`] = `
{
  "body": {
    "error": "validation",
    "validationErrors": [
      {
        "code": "too_small",
        "exact": false,
        "inclusive": true,
        "message": "String must contain at least 5 character(s)",
        "minimum": 5,
        "path": [
          "password",
        ],
        "type": "string",
      },
    ],
  },
  "headers": {
    "connection": "close",
    "content-length": "199",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`requests /auth/register should validate 4`] = `
{
  "body": {
    "error": "user exists",
  },
  "headers": {
    "connection": "close",
    "content-length": "23",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 400,
}
`;

exports[`requests /auth/register should verify with secret token 1`] = `
{
  "body": {
    "ok": true,
  },
  "headers": {
    "connection": "close",
    "content-length": "11",
    "content-security-policy": "default-src 'self';base-uri 'self';block-all-mixed-content;font-src 'self' https: data:;form-action 'self';frame-ancestors 'self';img-src 'self' data:;object-src 'none';script-src 'self';script-src-attr 'none';style-src 'self' https: 'unsafe-inline';upgrade-insecure-requests",
    "content-type": "application/json; charset=utf-8",
    "cross-origin-embedder-policy": "require-corp",
    "cross-origin-opener-policy": "same-origin",
    "cross-origin-resource-policy": "same-origin",
    "expect-ct": "max-age=0",
    "origin-agent-cluster": "?1",
    "referrer-policy": "no-referrer",
    "strict-transport-security": "max-age=15552000; includeSubDomains",
    "vary": "Accept-Encoding",
    "x-content-type-options": "nosniff",
    "x-dns-prefetch-control": "off",
    "x-download-options": "noopen",
    "x-frame-options": "SAMEORIGIN",
    "x-permitted-cross-domain-policies": "none",
    "x-xss-protection": "0",
  },
  "status": 200,
}
`;

exports[`requests /auth/register should verify with secret token 2`] = `[]`;
